src/App.tsx

```diff
@@ -1,26 +1,235 @@
+    setEditingTodoId(id);
+    try {
+      const editedTodo = await editTodo(id, data);
+
+      if (editedTodo.id) {
+        setTodos(prevTodos =>
+          prevTodos.map(todo => {
+            if (todo.id === id) {
+              return { ...editedTodo };
+            }
+
+            return todo;
+          }),
+        );
+        setEditingTodoId(null);
+      }
+    } catch {
+      setErrorMessage('Unable to update a todo');
+      setIsLoading(false);
+      throw new Error();
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const handleToggleAll = () => {
+    setIsLoading(true);
+    const allCompleted = todos.every(todo => todo.completed);
+    const newStatus = !allCompleted;
+
+    const todosToUpdate = todos.filter(todo => todo.completed !== newStatus);
+
+    Promise.allSettled(
+      todosToUpdate.map(todo =>
+        editTodo(todo.id, { completed: newStatus }).then(() => todo),
+      ),
+    )
+      .then(results => {
+        results.forEach(result => {
+          if (result.status === 'rejected') {
+            setErrorMessage('Unable to update a todo');
+          } else {
+            const updatedTodo = result.value as Todo;
+
+            setTodos(prevTodos =>
+              prevTodos.map(todo =>
+                todo.id === updatedTodo.id
+                  ? { ...todo, completed: newStatus }
+                  : todo,
+              ),
+            );
+          }
+        });
+      })
+      .finally(() => {
+        setIsLoading(false);
+      });
+  };
+
+  const filteredTodos = useMemo(
+    () => getPrepearedTodos(todos, filterCompleting),
+    [todos, filterCompleting],
+  );
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">
-          React Todo App - Add and Delete
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <Header
+          todos={todos}
+          tempTodo={tempTodo}
+          onAdd={handleAddTodo}
+          onError={handleError}
+          isError={!!errorMessage}
+          isLoading={isLoading}
+          isAllCompleted={todos.every(todo => todo.completed)}
+          onToggleAll={handleToggleAll}
+          ref={inputRef}
+        />
+
+        <TodoList
+          todos={filteredTodos}
+          tempTodo={tempTodo}
+          onDelete={handleDeleteTodo}
+          isLoading={isLoading}
+          deletingTodoId={deletingTodoId}
+          onEdit={handleEditTodo}
+          editingTodoId={editingTodoId}
+        />
+        {!!todos.length && (
+          <Footer
+            todos={todos.filter(todo => !todo.completed)}
```

Comment:
move to variable

Reply:
Even if i only use it once?

Reply:
it is better to get used to putting everything into variables, now that you have a small program, yes, but in the future, you may want to improve this program